rfd/0056-buildbox-image-handling.md

```diff
@@ -65,26 +65,107 @@ This process meets several of the goals set above:
+
+This is in order to guarantee that the image is what we think it is. This is a developer
+affordance, as all this can be accomplished with hashes, but it's cumbersome and 
+unfriendly.
+
+#### 4. Images get built on a tag update and after code review & CI
+
+This imposes the same release mechanism on the tooling as on the main software. The idea 
+is to turn the process of spinning a new buildbox version into a rare(-r) event, rather 
+than happening on every merge.
+
+#### 5. Everything (including releases) uses the same set of images
+
+A release process should refernce an _exact_ buildbox image it will use and commit that
+to `git` prior to building the release. We should be able to see which buildbox was used
+for any given release just by examining the Teleport source.
+
+#### 6. Enforce strong validation of components on the buildbox as far as possible
+
+Currently we implicitly trust that _none_ of the following have been compromsed when
+constructing our buildbox:
+ - The Ubuntu debian package repositories
+ - The `go` package download package
+ - The `rustup` tool download package
+ - The `etcd` download package
+ - The Google `addlicense` repository
+ - The `libbpf` download site
+ - The Google Cloud Platform SDK download site
+ - The `golangci-lint` download site
+ - and a few others, but you get the idea...
+
+At minimum, we should create SHAs of known-good versions of the downloaded artefacts 
+and verify that they have not changed, to ensure that we are not surprosed by any 
+changes after the fact.
+
+#### 7. Buildbox source extracted from Teleport repo into it's own repo
+
+(This is controversial)
+
+The whole `build.assets` directory is a murky part of the Teleport repo that I don't
+think people understand, so removing it from the teleport repo and having appropriate
+READMEs & so forth wouldbring it out into the open
+
+While this would require a 2-step shuffle where any buildbox image is built and released 
+prior to use by CI and Releases, this is also true for _any_  scenario that requires the 
+use of pre-built images for CI, as you can't build the image internally and then run a 
+build on it without requiring DinD or the like.
+
+_Author's Note:_ This is how I've managed buildbox vs. project code it in the past, so
+it feels natural to me. Reasonable people can differ on this,  and I accept that this
+rationale may not scale to the rest of the team. 😉
+
+### Open Questions
+
+#### Can we use Nix to do this for us?
+Don't know yet. A quick experiment suggests that using `nix` _would_ help repeatability 
+and automation, **_but_**:
+
+ * The generated images were _massive_ (~5GB for only a few of our dependencies), and
+ * The selection of versions available for any given package at a given time is quite 
+   limited. This may be fixable depensing on the version of `nixpkgs` we pin for any 
+   given branch, but it's a complication that needs to be consider.
```

Comment:
Anecdotally, nix is great for reproducibility and the folks who are working on the build directly.  However it can to be a major stumbling block for engineers who only peripherally interact with the build -- I think nix would only be worth investing in if we have a stable tools team with a long term vision of following the nix path.  Without that internal confidence, I wouldn't recommend introducing it partway in this RFD.

For further discussion https://news.ycombinator.com/item?id=22878155

FWIW, I think nix is one of the few systems that gets build determinism and (mostly) reproducibility correct.